Fabric DataFixer API [Javadoc Corner]

i509VCB opened the pull request:
Adds new API module fabric-datafixer-v1
Adds a SimpleFixes class full of common datafixes such as renaming blocks, items, entities, biomes.
Adds a DataFixerHelper method to help register the DataFixer.

### TODO:
Fully Implement saving mechanism to store mod's DataVersion within world save data, ditermine where and where this isn't necessary.

Test mod to verify DataFixing works. (Blocks and items work). Will be integrated into Test mods near release.

Tutorial page on wiki (The documentation Mojang should have written). 
Basically two tutorials:

How to use the simple DataFixes.
How to make your own DataFixes.

### WIP:

Writing Wiki Articles: 
A draft is here, will not be final: https://gist.github.com/i509VCB/f31ceab03730ecd8f936451d52ace810

Fixing (Block)Entities. Schema system is done, just need to verify it works and test it.

modmuss50 committed:
Fix builtin mods being added as resource packs

modmuss50 committed:
Update versioning to match 1.15 branch

modmuss50 committed:
Update fabric-loader to replace deprecated methods

Prospector committed:
Add more metadata to the modules (#353)

* Add fabric-api-base as a dep to fabric-networking-v0

parly committed:
Allow for running fabric-api-base on the server-side environment (#360)

modmuss50 committed:
Set curse version to 1.14.4
Also forces it to build again

Virtuoel committed:
Fix NPE in fluid renderer mixin (#361)

* Fix NPE in fluid renderer mixin
* Ensure state, view, and pos are never null
* Bump fabric-rendering-fluids-v1 version

i509VCB committed:
Skeleton Draft for DataFixer API

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/api/datafixer/FabricDataFixerUtils.java:
```diff
@@ -0,0 +1,98 @@
+    public static final DataFixer create(final String modid, final int runtimeDataVersion, final Consumer<DataFixerBuilder> builder) {
+        // Validation
+        Validate.notNull(modid, "modid cannot be null");
+        Validate.finite(runtimeDataVersion, "dataVersion must be finite");
```
What... this `Validate.finite` is for checking double/float NaNs.

Also we tend to use guava preconditions over apache validate.

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/api/datafixer/FabricDataFixerUtils.java:
```diff
@@ -0,0 +1,98 @@
+     * @param fix The datafix to add.
+     */
+    public static void addDataFix(final DataFixerBuilder builder, final DataFix fix) {
+        builder.addFixer(fix);
```
Uhh, why? people can just call the original method.

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/api/datafixer/FabricDataFixerUtils.java:
```diff
@@ -0,0 +1,98 @@
+    }
+    
+    @Deprecated
+    public static void test() {
```
Move these into a separate test source set.

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/api/datafixer/FabricDataFixerUtils.java:
```diff
@@ -0,0 +1,98 @@
+            // Make your schema with version to register the fixes under
+            Schema schema_1 = builder_1.addSchema(1, IDENTIFIER_NORMALIZE);
+            // Add your fix to schema you wish application to occur in.
+            SimpleFixes.addDataFixRenameBlock(builder_1, "rename example block", new ImmutableMap.Builder<String, String>().put("test:example1", "test:example").build(), schema_1);
```
`ImmutableMap.builder()`

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/api/datafixer/SimpleFixes.java:
```diff
@@ -0,0 +1,130 @@
+     * @param changes A map of all changed values, where the key should be original and value is new value.
+     * @param schema_1 The Schema to apply this fix to.
+     */
+    public static void addDataFixRenameBlock(DataFixerBuilder builder_1, String name, ImmutableMap<String, String> changes, Schema schema_1) {
```
`addBlockRenameFix` etc

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/api/datafixer/SimpleFixes.java:
```diff
@@ -0,0 +1,130 @@
+     * @param schema_1 The Schema to add this fix to.
+     */
+    public static void addDataFixRenameBiome(DataFixerBuilder builder_1, String name, ImmutableMap<String, String> changes, Schema schema_1) {
+        builder_1.addFixer(new DataFix(schema_1, false) {
```
I suggest moving this impl to a package private or a private subclass instead.

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/datafixer/mixin/util/SchemasAccessor.java:
```diff
@@ -0,0 +1,34 @@
+import net.minecraft.datafixers.Schemas;
+
+@Mixin(Schemas.class)
+public class SchemasAccessor {
```
Just use accessor mixins

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInvoker.java:
```diff
@@ -0,0 +1,84 @@
+
+public class FabricDataFixerInvoker {
+    
+    Map<String, DataFixerEntry> MOD_FIXERS = new HashMap<String, DataFixerEntry>();
```
Use diamond operator. It's java 8, not java 6.

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInvoker.java:
```diff
@@ -0,0 +1,84 @@
+    
+    private FabricDataFixerInvoker() {}
+    
+    public static final FabricDataFixerInvoker INSTANCE = new FabricDataFixerInvoker();;
```
extra semicolon

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInvoker.java:
```diff
@@ -0,0 +1,84 @@
+        Iterator<Entry<String, DataFixerEntry>> iterator = INSTANCE.MOD_FIXERS.entrySet().iterator();
+        CompoundTag currentTag = compoundTag_1;
+        
+        while(iterator.hasNext()) {
```
Use foreach loop

i509VCB replied:
Impossible because dataFixer.update(args...) returns a new CompoundTag and I have to replace local var with new tag as you see further down in iterator. If I were to use a foreach loop the compiler would fail because it wouldn't be final or effectively final.

Juuxel replied:
An actual foreach loop like this should not require final variables:
```java
for (Entry<String, DataFixerEntry> entry : INSTANCE.MOD_FIXERS.entrySet()) { ... }
```

i509VCB replied:
Yeah now I see, I was thinking of the lambda forEach stream loop and originally tried to avoid it because of streams performance.

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInvoker.java:
```diff
@@ -0,0 +1,84 @@
+    public static void addFixerVersions(CompoundTag compoundTag_1) {
+        Iterator<Entry<String, DataFixerEntry>> iterator = INSTANCE.MOD_FIXERS.entrySet().iterator();
+        
+        while(iterator.hasNext()) {
```
Foreach loop

liach commented on fabric-datafixer-v0/src/main/resources/server.mixins.json:
```diff
@@ -0,0 +1,11 @@
+  "package": "net.fabricmc.fabric.datafixer.mixin",
+  "compatibilityLevel": "JAVA_8",
+  "mixins": [
+    "MixinChunkSerializer"
```
Remember to add other mixin classes here!

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInvoker.java:
```diff
@@ -0,0 +1,84 @@
+        public DataFixer modFixer;
+        public int runtimeDataVersion;
+        
+        public DataFixerEntry(DataFixer fix, int runtimeDataVersion) {
```
Everything package private here

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInvoker.java:
```diff
@@ -0,0 +1,84 @@
+
+    class DataFixerEntry {
+        public DataFixer modFixer;
+        public int runtimeDataVersion;
```
final

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInvoker.java:
```diff
@@ -0,0 +1,84 @@
+import net.minecraft.util.crash.CrashReport;
+import net.minecraft.util.crash.CrashReportSection;
+
+public class FabricDataFixerInvoker {
```
make the class final

liach commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/datafixer/mixin/MixinTagHelper.java:
```diff
@@ -0,0 +1,24 @@
+public class MixinTagHelper {
+    @Inject(at = @At("RETURN"), method = "update(Lcom/mojang/datafixers/DataFixer;Lnet/minecraft/datafixers/DataFixTypes;Lnet/minecraft/nbt/CompoundTag;II)Lnet/minecraft/nbt/CompoundTag;")
+    private static void updateModFixers(DataFixer dataFixer_1, DataFixTypes dataFixTypes_1, CompoundTag compoundTag_1, int dynamicDataVersion, int runtimeDataVersion, CallbackInfoReturnable<CompoundTag> cir) {
+        CompoundTag normal = cir.getReturnValue(); // We do our fixes after vanilla.
```
Hmm, I think you capture the return values from the local vars instead of the cir

i509VCB replied:
Well TagHelper just wraps the whole method behind it's return statement and since I am applying my changes last I can just abuse the return value at end.

liach commented:
This draft on having per-mod data fixer instance is very nice. I am looking forward to another review when the pull request is feature complete (i.e. the code totally works)

i509VCB committed:
Draft PR v1.1

Replace iterators with nicer loops.
Renamed FabricDataFixerInvoker to FabricDataFixerImpl
Banished DataFixer creation inside of SimpleFixes to FabricSimpleFixesImpl

Did a few more reccomendations.

TODO:

Add BlockEntity rename fix.
Create TestMod within another repo or in ./testmods

i509VCB commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:
```diff
@@ -0,0 +1,82 @@
+    public DataFixer registerFixer(String modid, int runtimeDataVersion, DataFixer datafixer) {
+        
+        Preconditions.checkNotNull(modid, "modid cannot be null");
+        Preconditions.checkArgument(runtimeDataVersion > -1, "dataVersion must be finite");
```
Maybe a bit clickbaity, might want to change the argument again.

modmuss50 commented on fabric-datafixer-v0/build.gradle:
```diff
@@ -0,0 +1,7 @@
+archivesBaseName = "fabric-datafixer-v0"
```
New modules should be v1

modmuss50 commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/api/datafixer/FabricDataFixerUtils.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public class FabricDataFixerUtils {
+    
+    static boolean LOCKED = false;
```
I would just expose a getter to this in the API.

I know it doesnt stop someone from changing it, but it makes it a bit harder to do by accident.

modmuss50 commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/api/datafixer/SimpleFixes.java:
```diff
@@ -0,0 +1,96 @@
+     * @param changes A map of all changed values, where the key should be original and value is new value.
+     * @param schema_1 The Schema to apply this fix to.
+     */
+    public static void addBlockRenameFix(DataFixerBuilder builder_1, String name, ImmutableMap<String, String> changes, Schema schema_1) {
```
Drop the _1 from the param names. just `builder` and `schema`. I guess you got the idea for these from the decompiler.

modmuss50 commented on fabric-datafixer-v0/src/main/java/net/fabricmc/fabric/api/datafixer/SimpleFixes.java:
```diff
@@ -0,0 +1,96 @@
+        INSTANCE._addTransformEntityFix(builder_1, name, transformation, schema_1);
+    }
+    
+    protected abstract void _addTransformEntityFix(DataFixerBuilder builder_1, String name, EntityTransformation transformation, Schema schema_1);
```
Why have these methods here? The structure of this api seems a little odd?

i509VCB replied:
An earlier recommendation said to keep the actual datafixes I would be adding inside an impl class rather than leave the datafix mess inside api meant to be exposed to developers.

modmuss50 reviewed (changes requested):
Taken a quick look at this, looks good so far will be nice to see it when things are done some more 👍 

Just left a few points of basic feedback, let me know if you have any querys.

Draylar committed:
Cooler dimension API, #309 (#319)

* pyro API

* Update fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/EntityPlacer.java

clarify portalDir param

Co-Authored-By: Pyrofab <redstoneinfire@gmail.com>

* Update fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/EntityPlacer.java

rename & clarify verticalOffset param

Co-Authored-By: Pyrofab <redstoneinfire@gmail.com>

* Update fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/FabricDimensions.java

Co-Authored-By: Pyrofab <redstoneinfire@gmail.com>

* Update fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/EntityPlacer.java

Co-Authored-By: Pyrofab <redstoneinfire@gmail.com>

* update parameter

* Add fabric-networking and fabric-registry-sync to dependencies

* Refactor FabricDimensionType to use a builder

Also adds a desiredRawId field separate from the actual raw
id fixed internally. The fixedRawId field is now set by
DimensionIdsFixer through reflection.
This change addresses concerns about the ability for any mod
to set a FabricDimensionType's raw id at any time.

* Improve javadoc of FabricDimensions#teleport

* Add an overload for FabricDimensions#teleport

* Update javadoc on DimensionIdsFixer

* Inline clientside packet handling

* Add side assertions to FabricDimensions#teleport

* Nuke minecraft's attempts to overwrite bad level properties

* Add license headers

* Reformat dimension API source code

* Update FabricDimensionType javadoc

* Remove redundant ThreadLocal from FabricDimensionInternals

* Fix crash on dedicated servers

* Fix Illegal Access during remapping

* Fix dimension remap exception propagation

* clarify EntityPlacer docs

* spacing fixes

* api.dimension -> api.dimension.v1

grondag committed:
Increment API patch version

Signifies inclusion of new Dimension API

i509VCB committed:
Properly banish DataFixes to impl class.

Added temporary method to prepare for impl of renaming BlockEntities

Removed Mixins:
Removed LevelProperties (Injection covered by TagHelper)
Removed LevelStorage (Injection covered by TagHelper)
Removed PersistentStateManager (Injection covered by TagHelper and normal PersistentState)
Removed StructureManager (Injection covered by TagHelper)
Removed WorldSaveHandler (Injection covered by TagHelper)

i509VCB committed:
Move everything from v0 to v1 module.

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/SimpleFixes.java:
```diff
@@ -0,0 +1,90 @@
+/**
+ * This class file contains several common datafixes most modders would use.
+ */
+public abstract class SimpleFixes {
```
Can we just make this an interface given all the fields are public static final and all the methods are public and abstract?

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/datafixer/mixin/MixinPersistentState.java:
```diff
@@ -0,0 +1,10 @@
+
+@Mixin(PersistentState.class)
+public class MixinPersistentState {
+    // Add our mod's data versions to method_17919
```
Imo the mod data version should go into each persistent state instance, as each persistence state shouldreally be touched by only one mod.

i509VCB replied:
I looked around and saw these are the classes that extend the PersistentState (Excluding the Manager):

ChunkUpdateState
ForcedChunkState
RaidManager
ScoreboardState
MapState
IdCountsState

The whole reason I initially added it into method_17919 (closer to writeStateToFile in functionality) is to keep all the DataVersions together in the same spot. This is also for the implementation because it will read the mod provided DataVersions in CompoundTag to tell the DataFixer what Schema version the data it is looking at is.

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/datafixer/mixin/MixinPlayerEntity.java:
```diff
@@ -0,0 +1,11 @@
+
+@Mixin(PlayerEntity.class)
+public class MixinPlayerEntity {
+    // Add Mod DataVersions to writeCustomDataToTag
```
this sounds too expensive and will just blow up tags. can we just write ones we need?

i509VCB replied:
If you are asking, we only write the section for datafixers which have been registered. So if only two mods register fixers then we only add two entries to DataVersion list.

If we are concerned about larger than expected CompoundTags then that is a concern that will require more discussion.

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:
```diff
@@ -0,0 +1,82 @@
+        return datafixer;
+    }
+    
+    public static CompoundTag updateWithAllFixers(DataFixer dataFixer_1, DataFixTypes dataFixTypes_1, CompoundTag compoundTag_1, int dyanamicDataVersion, int runtimeDataVersion) {
```
Make it an instance method

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:
```diff
@@ -0,0 +1,82 @@
+        return currentTag;
+    }
+
+    public static void addFixerVersions(CompoundTag compoundTag_1) {        
```
Make it instance method

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:
```diff
@@ -0,0 +1,82 @@
+
+public final class FabricDataFixerImpl {
+    
+    Map<String, DataFixerEntry> MOD_FIXERS = Maps.newHashMap();
```
You should declare like
```java
    private final Map<String, DataFixerEntry> modFixers = new HashMap<>();
```

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricSimpleFixesImpl.java:
```diff
@@ -0,0 +1,103 @@
+import net.minecraft.datafixers.fixes.EntitySimpleTransformFix;
+import net.minecraft.datafixers.fixes.FixItemName;
+
+public class FabricSimpleFixesImpl extends SimpleFixes {
```
`FabricSimpleFixes`

i509VCB committed:
Banish code to instances and actually rename everything to v1 modules.

i509VCB committed:
Changes Part 3:
MixinServerStatHandler not needed because no logic uses asString

StructureValidatorProvider probably isn't needed for a Mixin because it just adds generic DataVersion if it does not contain it. Discussion needed

VersionedChunkStorage needs further discussion of weather it is nessecary for adding Mod DataVersion into tag or not.

i509VCB committed:
Accidentally forgot rest of commit

i509VCB commented:
VersionedChunkStorage needs further discussion of weather it is necessary to even add Mod DataVersions to.

i509VCB committed:
Another day, another patch.
Added a Custom Fabric Schema implementation that registers all ReferenceTypes.
Bug needing to be fixed:

[23:46:27] [main/ERROR]: Something went wrong upgrading!
java.util.NoSuchElementException: null
 at it.unimi.dsi.fastutil.ints.IntAVLTreeSet.firstInt(IntAVLTreeSet.java:942) ~[fastutil-8.2.1.jar:?]
 at com.mojang.datafixers.DataFixerUpper.getLowestFixSameVersion(DataFixerUpper.java:115) ~[datafixerupper-2.0.24.jar:?]
 at com.mojang.datafixers.DataFixerUpper.getRule(DataFixerUpper.java:127) ~[datafixerupper-2.0.24.jar:?]
 at com.mojang.datafixers.DataFixerUpper.update(DataFixerUpper.java:84) [datafixerupper-2.0.24.jar:?]
 at net.fabricmc.fabric.impl.datafixer.FabricDataFixerImpl.updateWithAllFixers(FabricDataFixerImpl.java:73) [fabric-datafixer-v1-0.1.0+fd9b895342-dev.jar:?]
 at net.minecraft.util.TagHelper.handler$updateModFixers$zcm000(TagHelper.java:568) [minecraft-1.14.4-mapped-net.fabricmc.yarn-1.jar:?]
 at net.minecraft.util.TagHelper.update(TagHelper.java:494) [minecraft-1.14.4-mapped-net.fabricmc.yarn-1.jar:?]
 at net.minecraft.util.TagHelper.update(TagHelper.java:490) [minecraft-1.14.4-mapped-net.fabricmc.yarn-1.jar:?]
 at net.minecraft.client.options.GameOptions.method_1626(GameOptions.java:486) [minecraft-1.14.4-mapped-net.fabricmc.yarn-1.jar:?]
 at net.minecraft.client.options.GameOptions.load(GameOptions.java:249) [minecraft-1.14.4-mapped-net.fabricmc.yarn-1.jar:?]
 at net.minecraft.client.options.GameOptions.<init>(GameOptions.java:211) [minecraft-1.14.4-mapped-net.fabricmc.yarn-1.jar:?]
 at net.minecraft.client.MinecraftClient.init(MinecraftClient.java:440) [minecraft-1.14.4-mapped-net.fabricmc.yarn-1.jar:?]
 at net.minecraft.client.MinecraftClient.start(MinecraftClient.java:394) [minecraft-1.14.4-mapped-net.fabricmc.yarn-1.jar:?]
 at net.minecraft.client.main.Main.main(Main.java:155) [minecraft-1.14.4-mapped-net.fabricmc.yarn-1.jar:?]
 at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
 at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
 at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
 at net.fabricmc.loader.game.MinecraftGameProvider.launch(MinecraftGameProvider.java:178) [net.fabricmc.yarn.1_14_4.1.net.fabricmc.fabric-loader-0.6.1+build.164.jar:?]
 at net.fabricmc.loader.launch.knot.Knot.init(Knot.java:131) [net.fabricmc.yarn.1_14_4.1.net.fabricmc.fabric-loader-0.6.1+build.164.jar:?]
 at net.fabricmc.loader.launch.knot.KnotClient.main(KnotClient.java:26) [net.fabricmc.yarn.1_14_4.1.net.fabricmc.fabric-loader-0.6.1+build.164.jar:?]

i509VCB committed:
First Success of functioning DataFix with renaming a block.

Still not complete for release, namely the hobbled mess of test code.

Also found a bug by accident in DataFixer.

Change where default mod DataVersion is 0 if no tag is present.

i509VCB committed:
Refactor and start draft of BlockEntityRenameFix.

i509VCB committed:
Remove uncatchable Crash Hook.

i509VCB committed:
Move BiomeRenameFix and BlockEntityFix to their own package rather than having anoymous classes.

i509VCB committed:
Added All Entities and BlockEntities to Schema.

Validated that EntityTransformation fix does work.

Added the extra 10+ Accessors needed to make Schema work.

i509VCB committed:
Add License headers

i509VCB committed:
Tab absolutely everything and make packages correct.

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/DataFixerUtils.java:
```diff
@@ -0,0 +1,68 @@
+ * <b>Please take extreme caution when using these tools as DataFixers directly interface with the world saves and may corrupt world saves.</b>
+ * </p>
+ * 
+ * @disclaimer <b>I, i509VCB and the Fabric Team do not take an responsibility for any worlds that are altered, destroyed, corrupted or damaged through the use of the {@link DataFixerUtils} because of improper configuration, mispellings, incorrect syntax, etc.</b>
```
:sweat: Isn't this already included in apache v2 license?

Juuxel commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/DataFixerUtils.java:
```diff
@@ -0,0 +1,68 @@
+	 */
+	public int getModDataVersion(CompoundTag compoundTag, String modid);
+
+	public boolean isLocked();
```
What does the return value mean?

liach replied:
guess it just prevents registering more fixers

Juuxel commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/SimpleFixes.java:
```diff
@@ -0,0 +1,115 @@
+	 * @param changes A map of all changed values, where the key should be original and value is new value.
+	 * @param schema The Schema to apply this fix to.
+	 */
+	public abstract void addBlockRenameFix(DataFixerBuilder builder_1, String name, String oldId, String newId, Schema schema_1);
```
The javadoc is missing oldId and newId.

liach replied:
Consider fixing the `_#` suffixes of parameters.

i509VCB committed:
Trying to reduce size of FabricSchema through several means.

i509VCB committed:
Reduced size of PR massively.

Figured out I needed to get proper key to grab MC Data Schemas.

TODO: Automatically resolve Schema version so Maintinence isn't needed.

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/DataFixerConstants.java:
```diff
@@ -0,0 +1,11 @@
+	/**
+	 * Represents the version of the latest vanilla schema version. This should be updated whenever a new Schema version is added to Schemas
+	 */
+	public static final int VANILLA_DATAFIXER_VERSION = 1961; // 1.14.4
```
Wait, I forgot why you can't grab this number from GameVersion

i509VCB replied:
That is because the latest schema version is 1961 and the World Data Version in SharedConstants for 1.14.4 is 1973.

Also you have to do some logic to convert schema version to right format so datafixer actually finds its vanilla schema.

modmuss50 replied:
This isnt ideal at all. Fabric can run across versions, and its not something I want to be updating each week.

i509VCB replied:
It does seem the Vanilla Datafixer stores its schema versions and I could possibly just rip the latest version number out of the map and automatically use that with a reflection hack.

Pyrofab commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/DataFixerUtils.java:
```diff
@@ -0,0 +1,68 @@
+ * <b>Please take extreme caution when using these tools as DataFixers directly interface with the world saves and may corrupt world saves.</b>
+ * </p>
+ */
+public interface DataFixerUtils {
```
`Utils` makes me think of static methods, maybe `Helper` would be more adequate ?

Pyrofab commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/SimpleFixes.java:
```diff
@@ -0,0 +1,116 @@
+		 * @param dynamic The Dynamic object representing the entity.
+		 * @return A Pair which contains the entity's new name and the dynamic representing the entity.
+		 */
+		public Pair<String, Dynamic<?>> transform(String inputEntityName, Dynamic<Tag> dynamic);
```
What type does the Dynamic have ? Tag or Entity ? Can the entity name be null ?

i509VCB replied:
The Dynamic is a Tag, I probably should fix that assuming the underlying logic that invokes the EntityTransformation will accept the casting differences.

Pyrofab commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:
```diff
@@ -0,0 +1,106 @@
+
+	@Override
+	public int getModDataVersion(CompoundTag compoundTag, String modid) {
+		return compoundTag.containsKey(modid + "_DataVersion", NbtType.NUMBER) ? compoundTag.getInt(modid + "_DataVersion") : 0;
```
`compoundTag.getInt` does exactly the same thing as this whole line.

i509VCB replied:
So getInt will return 0 if no element is present? If it doesn't return 0 then it is an issue since the default data version is 0

liach replied:
yes it returns 0 for nonexistent tags

Pyrofab commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:
```diff
@@ -0,0 +1,106 @@
+
+	final class DataFixerEntry {
+		private DataFixer modFixer;
+		private int runtimeDataVersion;
```
These fields could probably be final.

i509VCB committed:
Fix typing on EntityTrasformation Dynamics (untested).

Need to test a custom entity soon to verify if fixes apply across DataFixers or if I need to implement TypeReferenceHelper completely.

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/DataFixerConstants.java:
```diff
@@ -1,8 +1,10 @@
+ * @deprecated Alternative method to automatically resolve schema is being tested
+ * <p>This class contains constant DataFixer values for the DataFixer API.
  */
+@Deprecated
```
Just remove this if you need as this api is still a draft

Juuxel commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/SimpleFixes.java:
```diff
@@ -0,0 +1,116 @@
+	 * @param newBEName the new desired name of the BlockEntity being renamed.
+	 * @param schema The Schema to add this fix to.
+	 */
+	public abstract void addBlockEntityRenameFix(DataFixerBuilder builder, String name, String originalBEName, String newBEName, Schema schema);
```
`originalName` and `newName`? The method already knows that it's a BE.

i509VCB committed:
Added Reflection hacks to automatically resolve latest MC Schema version from DFU so Modmuss won't suffer having to update a value weekly.

grondag committed:
Particle API, BlockEntitySerializable.synch() and various fixes (#391)

* Added fabric-particles-v1

* Create an example mod for creating particles

* Add a sync() method to BlockEntityClientSerializable

* Attempt to prevent CME due to off-thread block entity updates (#367)

* Bump API version

i509VCB committed:
Bit of javadoc.

Also pushing so I can clean install the OS

i509VCB committed:
Ton O' Javadoc and addition of DataFixer entrypoint to handle entity and block entity registration.

i509VCB commented:
Okay I haven't done a whole ton with DataFixers in the last 2 weeks but I have an idea now what I need to change to fix the multi DataFixer delimma.

i509VCB committed:
Redo logic for mods defining entities and block entities for a safer system. Untested with registered (block) entities but more likely to work tomorrow.

i509VCB committed:
Part 1 of Shared Type Schema method. (Block)Entities have not been tested yet.

i509VCB committed:
Simple DataFixerHelper javadoc, Finalizing classes that shouldn't be extended.

i509VCB committed:
Formatting I, the </p> side of the force.

Tested, BlockEntities now can be registered with a Type<?> in build() instead of null.

i509VCB committed:
Formatting II: Abstract and Meme loicence edition.

liach committed:
Allows generic event declaration

Signed-off-by: liach <liach@users.noreply.github.com>

Player3324 committed:
Version bump

i509VCB committed:
Add BlockEntityTransformation (Functional), Null checking in FabricSimpleFixes.

i509VCB committed:
Renaming BlockEntities is functional now.

i509VCB commented:
The base fixes have been tested, now it is time for javadoc and wiki.

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:
```diff
@@ -187,7 +187,7 @@ public void registerTypes(Schema schema, Map<String, Supplier<TypeTemplate>> ent
 	 * Represents Minecraft's Built in DataFixer.
 	 */
-	public static class MCDFU {
+	public static final class MCDFU {
```
`McDfu`

ashpieboop replied:
Why not McDataFixerUpper or MinecraftDataFixerUpper or VanillaDataFixerUpper or VanillaDataFixer or MinecraftDataFixer ?

Commit:
Add checkstyle, format existing code accordingly. (#413)

Player3324 committed:
Add checkstyle verification for package names, fix non-API packages.

i509VCB committed:
Rename VanillaDataFixer and add a tiny amount of javadoc. More Javadoc to come explaining what the fixes do in more detail.

liach commented on fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:
```diff
@@ -187,14 +187,14 @@ public void registerTypes(Schema schema, Map<String, Supplier<TypeTemplate>> ent
 	 * Represents Minecraft's Built in DataFixer.
 	 */
-	public static final class MCDFU {
+	public static final class VanillaDataFixer {
```
Name this `VanillaDataFixers` to indicate this is a utility class.

i509VCB committed:
More Javadoc.

i509VCB committed:
Merge branch '1.14' of https://github.com/FabricMC/fabric into datafixer

i509VCB committed:
DataFix the checkstyle into order.

i509VCB committed:
Update wiki article link and add some debug logging.

i509VCB commented:
Okay let me resubmit this PR as targetting 1.15 will take a few mins.

i509VCB commented:
Replaced by #440
